swarm/storage/types.go

```diff
@@ -285,3 +311,76 @@ func (c ChunkData) Size() int64 {
 func (c ChunkData) Data() []byte {
 	return c[8:]
 }
+
+func NoValidateChunk(hasher SwarmHash, key *Key, data []byte) bool {
+	return true
+}
+
+// Provides method for validation of content address in chunks
+// Holds the corresponding hasher to create the address
+type ContentAddressValidator struct {
+	hashMu sync.Mutex
+	Hasher SwarmHash
+}
+
+// Constructor
+func NewContentAddressValidator(hash SwarmHash) *ContentAddressValidator {
+	return &ContentAddressValidator{
+		Hasher: hash,
+	}
+}
+
+// Validate that the given key is a valid content address for the given data
+func (self *ContentAddressValidator) Validate(key Key, data []byte) bool {
+	self.hashMu.Lock()
+	defer self.hashMu.Unlock()
+	self.Hasher.Reset()
+	self.Hasher.Write(data)
+	hash := self.Hasher.Sum(nil)
+
+	if !bytes.Equal(hash, key[:]) {
+		log.Error("invalid content address", "expected", fmt.Sprintf("%x", hash), "have", fmt.Sprintf("%x", key))
+		return false
+	}
+	return true
+}
+
+type NoopValidator struct {
+}
+
+func (self *NoopValidator) Validate(key Key, data []byte) bool {
+	return true
+}
+
+// Provides method for validating any chunk type
+type ChunkValidator interface {
+	Validate(Key, []byte) bool
+}
+
+type SequentialValidator struct {
```

Comment:
shall we not have sequential validation as an implementation and just provide a sequence of validators
empty array would become the noop. this could eliminate Noop, Sequential etc simplify the code

Reply:
Then, if we have both validators we need to put something like this in `Localstore.Put`

```
var valid bool
for _, v := range Localstore.Validators {
    if v.Validate(key, data) {
        valid = true
        break
    }
}
if !valid {
    return errors.New("invalid")
}
```

I fail to see how this is simpler. But maybe it has a slightly smaller footprint as it's one less stack frame.

Reply:
I also agree with @zelig - same as the commit I suggested earlier - https://github.com/nonsense/go-ethereum/commit/1b597764beea555da709a703b3ba954ac56cf247. I'd argue that less structs in this case is simpler to reason about. I don't have a better argument, so the difference is somewhat marginal.

Reply:
> so the difference is somewhat marginal

Hm, so I see it one way, you another. It's a bit like poh-tah-to, poh-taa-to.

So what do we do in such cases; especially considering it's not even very important?

Reply:
My opinion is:

When we have an odd number of reviewers/participants (in this case 3 - you, me and @zelig) majority decides on the appropriate solution. I don't have strong feelings about which solution we are going to use in this case, but I do believe that the []Validator is simpler and better, although I don't have any more arguments than those I have already given.

If there are even number of reviewers or just 2 (author + one), then the author of the code should decide what to use when the difference is marginal.